Add option to suppress logging output

tonigi opened the pull request:
This suppresses all stdout prints, with the exception of warnings.

Two function in `lib.py` are now provided to deal with output:
* `info()`, replacing print() for normal printouts, which is quenched by `--no-print`
* `warn()`, for warnings, which is not

Both can be reimplemented with loggers or other means if desired.

orbeckst commented:
A step in the right direction but I would prefer using the standard logger system.

tonigi commented:
That is easy now that print() is gone, but do we want to use loggers for all stdout?  timestamps and such may make the output unreadable.

orbeckst commented:
Personally, I like loggers everywhere but I see your point, especially for the more casual end user. (I should also say that I have no say in if this gets merged, I am just another user who contributed a tiny little bit to make the code more interoperable and scriptable).

tonigi commented:
I've made a `logging` branch as well, which uses loggers everywhere, if it can help.

orbeckst commented:
Cool. Whichever of the two branches gets merged — I look forward to `--no-output` (or perhaps name the option `--quiet` because "no output" could be interpreted as "do not write *any* output including files").

tonigi commented:
My preference was to merging this branch. The option is called `--no-print` indeed (misnamed pull request edited). (A `--quiet` stub was already there, so i did not interfere with it.)

matvey83 commented on propka/coupled_groups.py:
```diff
@@ -287,7 +288,7 @@ def transfer_determinant(self, determinants1, determinants2, label1, label2):
     #
 
-    def tagged_print(self, tag, s, labels):
+    def tagged_dprint(self, tag, s, labels):
```
Since this method doesn't do any printing, can we rename it?

matvey83 commented on propka/lib.py:
```diff
@@ -171,6 +171,8 @@ def loadOptions(*args):
            help="not activated yet")
     parser.add_option("--verbosity",  dest="verbosity", action="store_const",
            help="level of printout - not activated yet")
+    parser.add_option("--no-print",  dest="no_print", action="store_true", default=False,
```
Instead of adding a new option (--no-print), how about we get the existing --verbosity option to work?   If set to "none" it would be equivalent of --no-print.  In the future we could add more functionality - such as only printing out errors/warnings.

matvey83 commented on propka/lib.py:
```diff
@@ -272,3 +279,16 @@ def writeFile(filename, lines):
+
+
+def warn(*args, **kargs):
+    """Behaves like print() - possibly to be replaced by a logger"""
```
I think we should just go ahead and do this now.  Instead of a global "no_print" we could have a global "logger" variable.   If options.no_print is set (or options.verbosity is "none"), we would set the logger level to CRITICAL.  If not, it would be set to INFO.

Then we can replace these 2 functions with:

def info(*args):
    logging.info(' '.join(map(str, args)))

def warn(*args):
    logging.warning(' '.join(map(str, args)))

matvey83 commented:
Hi guys,

Got a chance to review these changes, and they look good.  Made a few suggestions as comments in the code.

Thanks

tonigi commented:
Ok, thanks! So, if there is a preference for all the std. output to go through a logger and filtered by the verbosity levels, I'll be adding the corresponding commits to this pull request in the coming days.

I guess that explicitly setting the Formatter should take care of possible (user-defined) logging format variability.

tonigi commented:
I've added commits so that the all the output goes through a global logger. 

Now, there remain the explicit "verbose=True" parameters and checks, which I guess are against the logger spirit. I removed them in the "logging-devel" branch, but oddly the regression test fails for ASP 25 A of 1HPX, and I can't see the root cause (messed up unnamed arguments, or different python reordering of dicts?).

matvey83 commented on propka/lib.py:
```diff
@@ -272,3 +287,31 @@ def writeFile(filename, lines):
+        logger.log(level,l)
+
+
+def info_debug(*args, **kargs):
```
Lets rename this to debug(), for consistency with logger.debug()

matvey83 commented on propka/lib.py:
```diff
@@ -272,3 +287,31 @@ def writeFile(filename, lines):
+def info(*args, **kargs):
+    """Log a message. Level defaults to INFO unless overridden."""
+    level = kargs.pop("level",logging.INFO)
+    for l in _sprint(*args, **kargs):
```
How about we move these lines into a common shared private function, and have info() and other 3 logging functions call it?

matvey83 commented on propka/lib.py:
```diff
@@ -272,3 +287,31 @@ def writeFile(filename, lines):
+    info(*args, **kargs, level=logging.DEBUG)
+
+
+def info_warning(*args, **kargs):
```
Lets rename this to warning(), for consistency with logger.warning().

matvey83 commented on propka/lib.py:
```diff
@@ -272,3 +287,31 @@ def writeFile(filename, lines):
+    for l in _sprint(*args, **kargs):
+        logger.warning("Warning: "+l)
+
+def _sprint(*args, **kargs):
```
This feels hacky.   Could we have info(), warning(), and debug() functions accept just *args that we could join together, without printing to QString?   Why are **kwargs needed?

Alternatively we could update the "print" statements in the codebase to call these functions with a single string (some would need to be updated to format multiple args into a single string).

matvey83 commented on propka/protonate.py:
```diff
@@ -407,31 +408,26 @@ def set_bond_distance(self, a, element):
 
         return a
 
     def display(self,*text):
```
If you want, you could replace all calls to display() with debug(), and remove display() completely.

matvey83 commented:
I ran the tests in your branch, and I'm getting these failures:
```
  File "/Users/adzhigir/src/propka-3.1/propka/lib.py", line 299
    info(*args, **kargs, level=logging.DEBUG)
                       ^
SyntaxError: invalid syntax
```

I was able to fix all test failures by updating the logging functions to:
```
def info(*args):
    """Log a message. Level defaults to INFO unless overridden."""
    logger.info(" ".join(map(str, args)))

def info_debug(*args):
    """Log a message on the DEBUG level."""
    logger.debug(" ".join(map(str, args)))

def info_warning(*args, **kargs):
    """Log a WARN message"""
    logger.warning(" ".join(map(str, args)))
```

tonigi commented:
All changes suggested were performed. Can you please have a look at the logging-devel branch failures?

matvey83 commented:
Test failures I'm getting with your logging-devel branch are:
```
Traceback (most recent call last):
  File "/Users/adzhigir/src/propka-3.1/scripts/propka31.py", line 12, in <module>
    import propka.lib, propka.molecular_container
  File "/Users/adzhigir/src/propka-3.1/propka/lib.py", line 299
    info(*args, **kargs, level=logging.DEBUG)
                       ^
SyntaxError: invalid syntax
Traceback (most recent call last):
  File "/Users/adzhigir/src/propka-3.1/scripts/propka31.py", line 12, in <module>
    import propka.lib, propka.molecular_container
  File "/Users/adzhigir/src/propka-3.1/propka/lib.py", line 299
    info(*args, **kargs, level=logging.DEBUG)
                       ^
SyntaxError: invalid syntax
Traceback (most recent call last):
  File "/Users/adzhigir/src/propka-3.1/scripts/propka31.py", line 12, in <module>
    import propka.lib, propka.molecular_container
  File "/Users/adzhigir/src/propka-3.1/propka/lib.py", line 299
    info(*args, **kargs, level=logging.DEBUG)
                       ^
SyntaxError: invalid syntax
```


Again, I was able to fix all test failures by updating the logging functions to:
```
def info(*args):
    """Log a message. Level defaults to INFO unless overridden."""
    logger.info(" ".join(map(str, args)))

def info_debug(*args):
    """Log a message on the DEBUG level."""
    logger.debug(" ".join(map(str, args)))

def info_warning(*args, **kargs):
    """Log a WARN message"""
    logger.warning(" ".join(map(str, args)))
```


Then I hit another error:
```
======================================================================
ERROR: test_pka_calc (pkacalc_test.py:SystemTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/adzhigir/src/propka-3.1/Tests/pkacalc_test.py", line 67, in test_pka_calc
    self.Fail("Unexpected pKa values:\n" + error_header +
AttributeError: 'SystemTest' object has no attribute 'Fail'
```


And had to update self.Fail() to self.fail().  Now the failure is:
```
       Group  Expected  Calculated
   ASP  25 A      5.07        8.64
   ASP  25 B      9.28        5.70
```

It's not clear why it's failing - I don't see this failure in your master branch.  You'll need to do some debugging - or start with working master branch and make forward changes one-by-one and see which change breaks it.

BTW it's confusing this pull request is for your master branch, yet your actual work is in the logging-devel branch.

tonigi commented:
Erm, the actual work IS in the master branch. The logging-devel is for tests.

matvey83 commented on propka/lib.py:
```diff
@@ -272,3 +287,36 @@ def writeFile(filename, lines):
+        logger.log(level, l)
+
+
+def _logsplit(*args):
```
How about simplifying things a little, and changing this whole section to:
```
def _args_to_str(arg_list):
    return " ".join(map(str, arg_list))

def info(*args):
    """Log a message. Level defaults to INFO unless overridden."""
    logger.info(_args_to_str(args))

def debug(*args):
    """Log a message on the DEBUG level."""
    logger.debug(_args_to_str(args))

def warning(*args, **kargs):
    """Log a WARN message"""
    logger.warning(_args_to_str(args))
```

tonigi replied:
ok, i think the printout is no longer exactly the same anyway.

matvey83 commented on propka/lib.py:
```diff
@@ -162,15 +167,13 @@ def loadOptions(*args):
-    parser.add_option("--verbosity",  dest="verbosity", action="store_const",
-           help="level of printout - not activated yet")
+           help="set the parameter file [%default]")
+    parser.add_option("--verbosity", dest="verbosity", action="store_const", default=1,
```
I think we should remove the --verbosity option (it wasn't working previously anyway), since we already have -q and -z options that do the same thing (-q suppresses all logging - verbosity=0, and -z enables debug mode - verbosity=2).

matvey83 commented on Tests/runtest.py:
```diff
@@ -66,9 +66,11 @@
                 print(" %s should be: %s" % (value, expected_value))
                 sys.exit(1)
 
+        """
```
Why was this block of code commented out?

tonigi replied:
I wrote that code to check other output files. But python random sorting of dictionaries makes printouts irreproducible.

matvey83 replied:
I see.  Lets just remove this block completely, instead of pushing something that is commented out.

If you want to add this test, feel free to file a separate pull request - we can discuss there how to get it to work.  For example, we could use OrderedDict instead of dict wherever needed.

Other then this, the current diff looks great, thanks for this work.

matvey83 commented on propka/atom.py:
```diff
@@ -5,6 +5,7 @@
 import string, propka.lib, propka.group
 
 from . import hybrid36
+from propka.lib import info, warning
```
This line is not needed, as info() nor warning() is used in the atom.py module.

matvey83 commented on propka/determinant.py:
```diff
@@ -1,6 +1,7 @@
 
 from __future__ import division
 from __future__ import print_function
+from propka.lib import info, warning
```
This line is not needed, as info() nor warning() is used in this module.

matvey83 commented on propka/determinants.py:
```diff
@@ -7,6 +7,7 @@
 import propka.iterative, propka.lib, propka.vector_algebra
 import propka.calculations
 from propka.determinant import Determinant
+from propka.lib import info, warning
```
This line is not needed, as info() nor warning() is used in this module.

matvey83 commented:
Looking great - just found a few unused imports; and would be good to remove the redundant --verbosity option, and then we are ready to merge.  :)

matvey83 commented:
Looking great, thanks!   One last thing remaining is to squash all of these commits into a single commit.  Then we can merge it into master and have a clean history.

tonigi commented:
I'm closing this request and creating a squashed one.
